README.md

```diff
@@ -18,5 +19,8 @@ a best case input for your implementation look like, what does a worst case
 input look like? How would this complexity change if you generated permutations
 randomly without memory instead of systematically trying them?
 
-Describe your reasoning and the conclusion you've come to. Your reasoning is the
-most important part. Add your answer to this markdown file.
+My implementation has an average complexity of $\Theta(n^n)$. Each recursive level has a complexity of approximately $n T(n-1) + n*n! + n^2$. For $i=n-3$ levels of recursion, this results in an approximate time complexity of $n^n + n^n n! + n^n$ which is an element of $\Theta(n^n)$.
```

Comment:
Not sure how you're getting $n^n$. Are you doing more than checking all permutations?

Reply:
I calculated $\Theta(n^n)$ using the following analysis:

$T(n) = 1$ for $n \leq 1$.
$T(n) = n(T(n-1) + (n-1)!) = n T(n-1) + n!$ for recursive complexity.

This is because of the outer most loop (line 23) runs $n$ times. Within this loop, I recursively call my function on an array of size $n-1$ (line 25) and run the inner loop on all $(n-1)!$ sub-permutations (line 27). I am ignoring the if statement on line 30 in my recursive analysis, since it does not evaluate to true in any of the recursive calls.

At recursive level $i$, $T(n) = n^i T(n-i)+(n^{i-1}+n^{i-2}+\cdots+1)n!$. $i$ equals $n-1$ to reach the base case, so $T(n) = n^{n-1}+\frac{n^{n-1}-1}{n-1} n!$. I think this is asymptotically equivalent to $n^n$.

Would you prefer I refactor my implementation to try to make it more efficient, or at least easier to analyze?

Reply:
Not sure if your analysis is correct. Are you checking a particular permutation more than once? It doesn't look to me like you're doing that, so unless you're doing a lot lot lot of work for each permutation, you should get something closer to $n!$.